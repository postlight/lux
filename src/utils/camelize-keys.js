import { camelize } from 'inflection';

const { entries } = Object;
const { isArray } = Array;

export default function camelizeKeys(obj, deep = false) {
  return isArray(obj) ? obj.slice() : entries(obj)
    .reduce((result, [key, value]) => {
      if (deep && value && typeof value === 'object' && !isArray(obj)) {
        value = camelizeKeys(value, true);
      }

      return {
        ...result,
        [camelize(key.replace(/-/g, '_'), true)]: value
      };
    }, {});
}
