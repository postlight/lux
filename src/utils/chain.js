// @flow
import type { Chain } from '../interfaces';

/**
 * @private
 */
export default function chain<T>(source: T): Chain<T> {
  return {
    pipe<U>(handler: (value: T) => U): Chain<U> {
      return chain(handler(source));
    },

    value(): T {
      return source;
    }
  };
}
