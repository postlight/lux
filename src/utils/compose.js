// @flow

/**
 * @private
 */
export function tap<T>(input: T): T {
  console.log(input); // eslint-disable-line no-console
  return input;
}

/**
 * @private
 */
export function compose<T>(...funcs: Array<(input: T) => T>): (input: T) => T {
  return input => funcs.reduceRight((value, fn) => fn(value), input);
}

/**
 * @private
 */
export function composeAsync<T>(
  ...funcs: Array<(input: T) => T | Promise<T>>
): (input: T) => Promise<T> {
  return input => funcs.reduceRight(
    (value, fn) => Promise.resolve(value).then(fn),
    Promise.resolve(input)
  );
}
