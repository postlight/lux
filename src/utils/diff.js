// @flow

/**
 * @private
 */
export function map<K, V>(a: Map<K, V>, b: Map<K, V>): Map<K, V> {
  return Array
    .from(b)
    .reduce((result, [key, value]) => {
      if (a.get(key) !== value) {
        result.set(key, value);
      }

      return result;
    }, new Map());
}
