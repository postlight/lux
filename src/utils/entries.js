// @flow
const HAS_OBJECT_ENTRIES = typeof Object.entries === 'function';

/**
 * @private
 */
export default function entries(source: Object): Array<[string, any]> {
  if (HAS_OBJECT_ENTRIES) {
    return Object.entries(source);
  }

  return Object.keys(source).reduce((result, key) => {
    const value = Reflect.get(source, key);

    result.push([key, value]);

    return result;
  }, []);
}
