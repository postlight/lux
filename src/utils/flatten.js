/* @flow */

type TwoDimArray<T> = Array<T | Array<T>>
type MultiDimArray<T> = Array<$Supertype<T | MultiDimArray<T>>>

export const shallow = <T>(source: TwoDimArray<T>): Array<T> =>
  Array.prototype.concat.apply([], source)

export const deep = <T>(source: MultiDimArray<T>): Array<T> =>
  source.reduce((dest, node) => {
    let flatNode = node

    if (Array.isArray(flatNode)) {
      flatNode = deep(flatNode)
    }

    return dest.concat(flatNode)
  }, [])
