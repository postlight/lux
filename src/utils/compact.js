/* @flow */

import { isNil } from '@lux/utils/is-type'

const compact = <T>(source: Array<T>): Array<T> =>
  source.filter(value => !isNil(value))

export default compact
