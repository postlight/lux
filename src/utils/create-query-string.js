// @flow
import entries from './entries';

/**
 * A replacement for querystring.stringify that supports nested objects.
 *
 * @private
 */
export default function createQueryString(
  source: Object,
  property?: string
): string {
  return entries(source).reduce((result, [key, value], index, arr) => {
    if (index > 0) {
      result += '&';
    }

    if (property) {
      result += (
        property
        + encodeURIComponent('[')
        + key
        + encodeURIComponent(']')
        + '='
      );
    } else {
      result += `${key}=`;
    }

    if (value && typeof value === 'object') {
      if (Array.isArray(value)) {
        result += value.map(encodeURIComponent).join();
      } else {
        result = (
          result.substr(0, result.length - (key.length + 1))
          + createQueryString(value, key)
        );
      }
    } else if (!value && typeof value !== 'number') {
      result += 'null';
    } else {
      result += encodeURIComponent(value);
    }

    return result;
  }, '');
}
