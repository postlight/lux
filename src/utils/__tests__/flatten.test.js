/* @flow */

import * as flatten from '@lux/utils/flatten'

test('#deep()', () => {
  expect(flatten.deep([1, [2], [[3]]])).toMatchSnapshot()
})

test('#shallow()', () => {
  expect(flatten.shallow([1, [2], [[3]]])).toMatchSnapshot()
})
