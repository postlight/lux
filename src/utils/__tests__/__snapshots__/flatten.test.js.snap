// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#deep() 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`#shallow() 1`] = `
Array [
  1,
  2,
  Array [
    3,
  ],
]
`;
